aten/src/ATen/native/sparse/SparseFactories.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+#include <ATen/TensorIndexing.h>
+#include <ATen/TensorIterator.h>
+#include <ATen/core/ATen_fwd.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/native/TensorFactories.h>
+#include <c10/core/Scalar.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Exception.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/sparse_coo_tensor.h>
+#endif
+
+namespace at {
+namespace native {
+namespace impl {
+
+template <typename setup_func_t, typename kernel_func_t>
+Tensor spdiags_impl(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape,
+    c10::optional<Layout> layout,
+    const setup_func_t& setup_func,
+    const kernel_func_t& kernel_func) {
+  auto diagonals_2d = diagonals.dim() == 1 ? diagonals.unsqueeze(0) : diagonals;
+  TORCH_CHECK(diagonals_2d.dim() == 2, "Diagonals must be vector or matrix");
+  TORCH_CHECK(shape.size() == 2, "Output shape must be 2d");
+  auto offsets_1d = offsets.dim() == 0 ? offsets.unsqueeze(0) : offsets;
+  TORCH_CHECK(offsets_1d.dim() == 1, "Offsets must be scalar or vector");
+  TORCH_CHECK(
+      diagonals_2d.size(0) == offsets_1d.size(0),
+      "Number of diagonals (",
+      diagonals_2d.size(0),
+      ") does not match the number of offsets (",
+      offsets_1d.size(0),
+      ")");
+  if (layout) {
+    TORCH_CHECK(
+        (*layout == Layout::Sparse) || (*layout == Layout::SparseCsc) ||
+            (*layout == Layout::SparseCsr),
+        "Only output layouts (",
+        Layout::Sparse,
+        ", ",
+        Layout::SparseCsc,
+        ", and ",
+        Layout::SparseCsr,
+        ") are supported");
+  }
+  TORCH_CHECK(
+      offsets_1d.scalar_type() == at::kLong,
+      "spdiags(): Expected a LongTensor of offsets but got ",
+      offsets_1d.scalar_type());
+
+  auto nnz_per_diag = at::zeros_like(offsets_1d);
+
+  auto setup_iter = TensorIteratorConfig()
+                        .add_output(nnz_per_diag)
+                        .add_input(offsets_1d)
+                        .build();
+  int64_t nnz = 0;
+  // Checks offsets and computes nnz per diag
+  setup_func(setup_iter, shape[0], shape[1], diagonals_2d.size(1), nnz);
+  auto n_diag = offsets_1d.size(0);
+  auto nnz_prefix = nnz_per_diag.cumsum(0).sub(nnz_per_diag);
```

Comment:
```suggestion
  auto nnz_prefix = nnz_per_diag.cumsum(0).sub_(nnz_per_diag);
```

Reply:
Also,
```c++
auto nnz_cumsum = nnz_per_diag.cumsum(-1);
const auto nnz = nnz_cumsum.select(0, -1).item<int64_t>();
const auto nnz_prefix = nnz_cumsum.sub_(nnz_per_diag);
```
This resolves the race condition from above.

Reply:
I did this with a small modification. The `select(0, -1)` is out of bounds for an edge case where `n_diag` is zero.